reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/Debounce.kt

```diff
@@ -1,66 +1,7 @@
 package com.badoo.reaktive.observable
 
-import com.badoo.reaktive.base.subscribeSafe
-import com.badoo.reaktive.disposable.CompositeDisposable
-import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.completable.completableTimer
 import com.badoo.reaktive.scheduler.Scheduler
-import com.badoo.reaktive.utils.atomic.AtomicReference
-import com.badoo.reaktive.utils.atomic.update
 
 fun <T> Observable<T>.debounce(timeoutMillis: Long, scheduler: Scheduler): Observable<T> =
-    observable { emitter ->
-        val disposables = CompositeDisposable()
-        emitter.setDisposable(disposables)
-        val executor = scheduler.newExecutor()
-        disposables += executor
-
-        subscribeSafe(
-            object : ObservableObserver<T> {
-                private val pendingValue = AtomicReference<DebouncePendingValue<T>?>(null, true)
-
-                override fun onSubscribe(disposable: Disposable) {
-                    disposables += disposable
-                }
-
-                override fun onNext(value: T) {
-                    val newPendingValue = DebouncePendingValue(value)
-                    pendingValue.value = newPendingValue
-
-                    executor.cancel()
-
-                    executor.submit(timeoutMillis) {
-                        pendingValue.update {
-                            if (it === newPendingValue) null else it
-                        }
-
-                        emitter.onNext(value)
-                    }
-                }
-
-                override fun onComplete() {
-                    executor.cancel()
-
-                    executor.submit {
-                        pendingValue
-                            .value
-                            ?.value
-                            ?.also(emitter::onNext)
-
-                        emitter.onComplete()
-                    }
-                }
-
-                override fun onError(error: Throwable) {
-                    executor.cancel()
-
-                    executor.submit {
-                        emitter.onError(error)
-                    }
-                }
-            }
-        )
-    }
-
-private class DebouncePendingValue<T>(
-    val value: T
-)
\ No newline at end of file
+    debounce { completableTimer(timeoutMillis, scheduler) }
```

Comment:
This is much less efficient then it was as there is a Completable per every item and each creates an instance of Executor. Let's stick with two different implementations. Think about maybe it's possible to extract some common logic to a method or class.

Reply:
It seems that only `DebouncePendingValue` can be shared between two implementations because other common parts are too small to be extracted. So I made `DebouncePendingValue` internal instead of private.